op-node/rollup/clsync/clsync.go

```diff
@@ -58,67 +55,124 @@ func (eq *CLSync) LowestQueuedUnsafeBlock() eth.L2BlockRef {
 	return ref
 }
 
-// AddUnsafePayload schedules an execution payload to be processed, ahead of deriving it from L1.
-func (eq *CLSync) AddUnsafePayload(envelope *eth.ExecutionPayloadEnvelope) {
-	if envelope == nil {
-		eq.log.Warn("cannot add nil unsafe payload")
-		return
+type ReceivedUnsafePayloadEvent struct {
+	Envelope *eth.ExecutionPayloadEnvelope
+}
+
+func (ev ReceivedUnsafePayloadEvent) String() string {
+	return "received-unsafe-payload"
+}
+
+func (eq *CLSync) OnEvent(ev rollup.Event) {
+	// Events may be concurrent in the future. Prevent unsafe concurrent modifications to the payloads queue.
+	eq.mu.Lock()
+	defer eq.mu.Unlock()
+
+	switch x := ev.(type) {
+	case engine.InvalidPayloadEvent:
+		eq.onInvalidPayload(x)
+	case engine.ForkchoiceUpdateEvent:
+		eq.onForkchoiceUpdate(x)
+	case ReceivedUnsafePayloadEvent:
+		eq.onUnsafePayload(x)
 	}
+}
 
-	if err := eq.unsafePayloads.Push(envelope); err != nil {
-		eq.log.Warn("Could not add unsafe payload", "id", envelope.ExecutionPayload.ID(), "timestamp", uint64(envelope.ExecutionPayload.Timestamp), "err", err)
-		return
+// onInvalidPayload checks if the first next-up payload matches the invalid payload.
+// If so, the payload is dropped, to give the next payloads a try.
+func (eq *CLSync) onInvalidPayload(x engine.InvalidPayloadEvent) {
+	eq.log.Debug("CL sync received invalid-payload report", x.Envelope.ExecutionPayload.ID())
+
+	block := x.Envelope.ExecutionPayload
+	if peek := eq.unsafePayloads.Peek(); peek != nil &&
+		block.BlockHash == peek.ExecutionPayload.BlockHash {
+		eq.log.Warn("Dropping invalid unsafe payload",
+			"hash", block.BlockHash, "number", uint64(block.BlockNumber),
+			"timestamp", uint64(block.Timestamp))
+		eq.unsafePayloads.Pop()
 	}
-	p := eq.unsafePayloads.Peek()
-	eq.metrics.RecordUnsafePayloadsBuffer(uint64(eq.unsafePayloads.Len()), eq.unsafePayloads.MemSize(), p.ExecutionPayload.ID())
-	eq.log.Trace("Next unsafe payload to process", "next", p.ExecutionPayload.ID(), "timestamp", uint64(p.ExecutionPayload.Timestamp))
 }
 
-// Proceed dequeues the next applicable unsafe payload, if any, to apply to the tip of the chain.
-// EOF error means we can't process the next unsafe payload. The caller should then try a different form of syncing.
-func (eq *CLSync) Proceed(ctx context.Context) error {
+// onForkchoiceUpdate peeks at the next applicable unsafe payload, if any,
+// to apply on top of the received forkchoice pre-state.
+// The payload is held on to until the forkchoice changes (success case) or the payload is reported to be invalid.
+func (eq *CLSync) onForkchoiceUpdate(x engine.ForkchoiceUpdateEvent) {
```

Comment:
nit: At some point it would be good to rename `eq` here to something more indicative of `CLSync`.

Reply:
Yep, remnant of the engine-queue. We can clean up the struct method receiver names some time in a sweep over all structs. Maybe as part of the final events polish step :+1: